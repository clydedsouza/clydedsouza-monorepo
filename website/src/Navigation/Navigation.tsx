import { Outlet, Link } from "react-router-dom";

function Navigation() {
    return (
      <div>
       <nav>
        <ul>
          <li>
            <Link to="/">Home</Link>
          </li>
          <li>
            <Link to="/projects">Blogs</Link>
          </li>
          <li>
            <Link to="/headline">Healdine</Link>
          </li>
        </ul>
      </nav>

      <Outlet />
      </div>
    );
  }
  
  export default Navigation;