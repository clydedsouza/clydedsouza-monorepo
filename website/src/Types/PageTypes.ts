export enum PageTypes {
  About = 'About',
  Projects = 'Projects',
  Speaking = 'Speaking',
  Teaching = 'Teaching',
  Books = 'Books',
  Highlights = 'Highlights',
  Pinned = 'Pinned',
}
