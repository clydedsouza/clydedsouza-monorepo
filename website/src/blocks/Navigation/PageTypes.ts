export enum PageTypes {
  About = "About",
  Portfolio = "Portfolio",
  Platforms = "Platforms",
  Highlights = "Highlights",
  Cta = "Cta",
  Blog = "Blog",
}
