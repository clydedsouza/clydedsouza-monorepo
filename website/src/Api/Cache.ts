import { PageTypes } from '../Types/PageTypes'
import { IProjectData } from '../Types/ProjectData'
import { getProjectData } from './Core'

const projectDataCacheMap: Map<PageTypes, Promise<IProjectData>> = new Map()

export const getCachedProjectData = (
  pageType: PageTypes
): Promise<IProjectData> => {
  if (!projectDataCacheMap.has(pageType)) {
    const getProjectDataPromise = getProjectData(pageType).catch((error) => {
      projectDataCacheMap.delete(pageType)
      return Promise.reject(error)
    })
    projectDataCacheMap.set(pageType, getProjectDataPromise)
  }

  return projectDataCacheMap.get(pageType)!
}
