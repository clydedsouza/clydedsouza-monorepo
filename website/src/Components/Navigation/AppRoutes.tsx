import { BrowserRouter, NavLink, Outlet, Route, Routes } from 'react-router-dom'
import AboutMe from '../../Pages/AboutMe/AboutMe'
import Projects from '../../Pages/Projects/Projects'
import { PageTypes } from '../../Types/PageTypes'

function Navigation() {
  return (
    <>
      <nav>
        <NavLink
          to="/"
          className={({ isActive }) => (isActive ? 'active' : '')}
        >
          {PageTypes.About}
        </NavLink>
        <NavLink
          to="/highlights"
          className={({ isActive }) => (isActive ? 'active' : '')}
        >
          {PageTypes.Highlights}
        </NavLink>
        <NavLink
          to="/platforms"
          className={({ isActive }) => (isActive ? 'active' : '')}
        >
          {PageTypes.Platforms}
        </NavLink>
      </nav>
      <Outlet />
    </>
  )
}

function AppRoutes() {
  return (
    <BrowserRouter>
      <Routes>
        <Route path="/" element={<Navigation />}>
          <Route index element={<AboutMe />} />
          <Route
            path={PageTypes.Highlights.toLowerCase()}
            element={<Projects name={PageTypes.Highlights} />}
          />
          <Route
            path={PageTypes.Platforms.toLowerCase()}
            element={<Projects name={PageTypes.Platforms} />}
          />
          {/* <Route path="*" element={<NoPage />} /> */}
        </Route>
      </Routes>
    </BrowserRouter>
  )
}

export default AppRoutes
