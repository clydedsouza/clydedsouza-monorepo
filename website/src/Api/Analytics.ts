import mixpanel from 'mixpanel-browser'
import { PageTypes } from '../Components/Navigation/PageTypes'
import {
  ILinkClickedProps,
  MIXPANEL_DEV_PROJECT_ID,
  MIXPANEL_PROD_PROJECT_ID,
} from './IAnalytics'

export const initAnalyticsWithSuperProperties = () => {
  const mixpanelProjectId =
    process.env.NODE_ENV === 'production'
      ? MIXPANEL_PROD_PROJECT_ID
      : MIXPANEL_DEV_PROJECT_ID

  try {
    mixpanel.init(mixpanelProjectId)
    mixpanel.register({
      environment: process.env.NODE_ENV,
    })
  } catch {
    undefined
  }
}

export const sendPageViewEvent = (pageType: PageTypes) => {
  try {
    mixpanel.track('Page view', { page: pageType })
  } catch {
    undefined
  }
}

export const sendLinkClickedEvent = (props: ILinkClickedProps) => {
  try {
    mixpanel.track('Link clicked', {
      destination_url: props.link,
      link_type: props.type,
      page_location: props.location,
    })
  } catch {
    undefined
  }
}
