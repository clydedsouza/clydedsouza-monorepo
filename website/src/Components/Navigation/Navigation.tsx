import { BrowserRouter, NavLink, Outlet, Route, Routes } from 'react-router-dom'
import AboutMe from '../../Pages/AboutMe/AboutMe'
import Projects from '../../Pages/Projects/Projects'
import { PageTypes } from '../../Types/PageTypes'
import './Navigation.scss'

function Menu() {
  return (
    <>
      <nav>
        <NavLink
          to="/"
          className={({ isActive }) => (isActive ? 'active' : '')}
        >
          {PageTypes.About}
        </NavLink>
        <NavLink
          to={PageTypes.Highlights.toLowerCase()}
          className={({ isActive }) => (isActive ? 'active' : '')}
        >
          {PageTypes.Highlights}
        </NavLink>
        <NavLink
          to={PageTypes.Platforms.toLowerCase()}
          className={({ isActive }) => (isActive ? 'active' : '')}
        >
          {PageTypes.Platforms}
        </NavLink>
      </nav>
      <Outlet />
    </>
  )
}

function Navigation() {
  return (
    <BrowserRouter>
      <Routes>
        <Route path="/" element={<Menu />}>
          <Route index element={<AboutMe />} />
          <Route
            path={PageTypes.Highlights.toLowerCase()}
            element={<Projects name={PageTypes.Highlights} />}
          />
          <Route
            path={PageTypes.Platforms.toLowerCase()}
            element={<Projects name={PageTypes.Platforms} />}
          />
          <Route path="*" element={<AboutMe />} />
        </Route>
      </Routes>
    </BrowserRouter>
  )
}

export default Navigation
