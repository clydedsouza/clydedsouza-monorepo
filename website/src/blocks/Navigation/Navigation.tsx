"use client";

import { sendNavigationClickedEvent } from "@/api/Analytics";
import Link from "next/link";
import { usePathname } from "next/navigation";
import { useState } from "react";
import "./Navigation.scss";
import { PageTypes } from "./PageTypes";

const getCurrentPage = (pathFromNextJs: string) => {
  const cleanPath = pathFromNextJs.replaceAll("/", "").toLowerCase();

  if (cleanPath === PageTypes.Platforms.toLowerCase()) {
    return PageTypes.Platforms;
  } else if (cleanPath === PageTypes.Portfolio.toLowerCase()) {
    return PageTypes.Portfolio;
  } else {
    return PageTypes.About;
  }
};

function Navigation() {
  const [currentPage, setCurrentPage] = useState(getCurrentPage(usePathname()));

  const onNavigationItemClick = (pageType: PageTypes) => {
    sendNavigationClickedEvent(pageType);
    setCurrentPage(pageType);
  };

  return (
    <>
      <nav>
        <Link
          href="/"
          className={currentPage === PageTypes.About ? "active" : ""}
          onClick={() => onNavigationItemClick(PageTypes.About)}
        >
          {PageTypes.About}
        </Link>
        <Link
          href={PageTypes.Portfolio.toLowerCase()}
          className={currentPage === PageTypes.Portfolio ? "active" : ""}
          onClick={() => onNavigationItemClick(PageTypes.Portfolio)}
        >
          {PageTypes.Portfolio}
        </Link>
        <Link
          href={PageTypes.Platforms.toLowerCase()}
          className={currentPage === PageTypes.Platforms ? "active" : ""}
          onClick={() => onNavigationItemClick(PageTypes.Platforms)}
        >
          {PageTypes.Platforms}
        </Link>
        <Link
          href={"https://blog.clydedsouza.net"}
          target="_blank"
          onClick={() => sendNavigationClickedEvent(PageTypes.Blog)}
        >
          {PageTypes.Blog}
        </Link>
      </nav>
    </>
  );
}

export default Navigation;
