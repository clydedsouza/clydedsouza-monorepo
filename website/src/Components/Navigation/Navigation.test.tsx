import {
  render,
  screen,
  waitForElementToBeRemoved,
} from '@testing-library/react'
import * as apiCacheModule from '../../Api/Cache'
import Navigation from './Navigation'
import { PageTypes } from './PageTypes'

jest.mock('../../Api/Cache')

describe('Navigation', () => {
  beforeEach(() => {
    jest
      .spyOn(apiCacheModule, 'getCachedProjectData')
      .mockImplementationOnce(() =>
        Promise.resolve({
          app: {},
          data: [],
        })
      )
  })

  it('should render navigation', async () => {
    render(<Navigation />)
    await waitForElementToBeRemoved(() => screen.queryByRole('progressbar'))
    expect(document.body).toMatchSnapshot()
  })

  it('should call API with correct page type', async () => {
    render(<Navigation />)
    await waitForElementToBeRemoved(() => screen.queryByRole('progressbar'))
    expect(apiCacheModule.getCachedProjectData).toBeCalledTimes(1)
    expect(apiCacheModule.getCachedProjectData).toBeCalledWith(
      PageTypes.Highlights
    )
  })
})
