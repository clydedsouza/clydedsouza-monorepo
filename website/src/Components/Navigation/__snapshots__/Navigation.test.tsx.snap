// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation should render navigation 1`] = `
<body>
  <div>
    <nav>
      <a
        aria-current="page"
        class="active"
        href="/"
      >
        About
      </a>
      <a
        class=""
        href="/highlights"
      >
        Highlights
      </a>
      <a
        class=""
        href="/platforms"
      >
        Platforms
      </a>
    </nav>
    <h2>
      About me
    </h2>
    <p>
      Hi! I'm Clyde D'Souza. I'm a creative and self-driven person who likes to code and do other things. I'm currently working as a senior software engineer at
       
      <a
        href="https://www.xero.com/nz/?ref=clydedsouza.net"
        rel="noreferrer"
        target="_blank"
      >
        Xero
      </a>
      . Previously, I've worked with
       
      <a
        href="https://www.datacom.co.nz/?ref=clydedsouza.net"
        rel="noreferrer"
        target="_blank"
      >
        Datacom
      </a>
       
      and
       
      <a
        href="https://www.heritagehotels.co.nz/?ref=clydedsouza.net"
        rel="noreferrer"
        target="_blank"
      >
        Heritage Hotels
      </a>
      .
    </p>
    <p>
      I was the event director of a non-profit educational event called Light & Spark NPO. I conceptualized and executed this event in Mumbai along with my
       
      <a
        href="https://lightandsparknpo.github.io/team/"
        rel="noreferrer"
        target="_blank"
      >
        amazing team
      </a>
      . Read more about it
       
      <a
        href="https://lightandsparknpo.github.io/2019-11-20-ruia-mumbai-2018/"
        rel="noreferrer"
        target="_blank"
      >
        here
      </a>
      .
    </p>
    <p>
      I've published a book titled Mama, Tell Me a Story which is available on
       
      <a
        href="http://mamatellmeastory.clydedsouza.net/"
        rel="noreferrer"
        target="_blank"
      >
        these platforms
      </a>
      . It's a collection of twelve short bedtime stories that parents will love reading to their kids over and over again. Go buy it now!
    </p>
    <p>
      In my spare time, I like to explore other interests like
       
      <a
        href="https://medium.com/@clydedz"
        rel="noreferrer"
        target="_blank"
      >
        writing
      </a>
      ,
       
      <a
        href="https://www.behance.net/clydedz"
        rel="noreferrer"
        target="_blank"
      >
        designing
      </a>
      ,
       
      <a
        href="https://www.skillshare.com/r/user/clydedsouza"
        rel="noreferrer"
        target="_blank"
      >
        teaching
      </a>
      , and I've also created many
       
      <a
        href="/#/projects"
      >
        other projects
      </a>
      .
    </p>
    <p>
      I'm an alumnus of
       
      <a
        href="https://www.aut.ac.nz/?ref=clydedsouza"
        rel="noreferrer"
        target="_blank"
      >
        Auckland University of Technology (AUT)
      </a>
       
      with a Postgraduate Diploma in Computer and Information Sciences and holds a Bachelor of Science degree specializing in Information Technology from University of Mumbai, India.
    </p>
    <h2>
      Highlights
    </h2>
    <div
      id="pinnedProjects"
    >
      <div
        id="pinnedProjectsList"
      >
        <p>
          Couldnâ€˜t load the data at this stage
        </p>
      </div>
    </div>
  </div>
</body>
`;
